Depedendent context functions related to higher kinded types seem to make the compiler susceptible to crashes.

## Compiler version

3.8.2

## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

```Scala
trait A[B[_]]:
  type C[D]

class B[E]:
  def apply[D[_]](other: (a: A[D]) ?=> a.C[E]): B[E] = ???

object B:
  given A[B]

def b: B[Int] = ???

b(b)
```

## Output (click arrow to expand)
<!--
Note: to get a full crash log when using scala-cli, run compilation with the flag `--server=false`
-->
<details>

```scala
    java.lang.AssertionError: assertion failed: `wildApprox` failed to remove uninstantiated D
    	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:10)
    	at dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:684)
    	at dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:7073)
    	at dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:7073)
    	at dotty.tools.dotc.core.Types$TypeAccumulator.foldOver(Types.scala:7000)
    	at dotty.tools.dotc.core.Types$TypeTraverser.traverseChildren(Types.scala:7074)
    	at dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:692)
    	at dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:7073)
    	at dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:7073)
    	at dotty.tools.dotc.core.Types$TypeAccumulator.op$proxy28$1(Types.scala:6983)
    	at dotty.tools.dotc.core.Types$TypeAccumulator.foldArgs$3(Types.scala:6983)
    	at dotty.tools.dotc.core.Types$TypeAccumulator.foldOver(Types.scala:6987)
    	at dotty.tools.dotc.core.Types$TypeTraverser.traverseChildren(Types.scala:7074)
    	at dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:692)
    	at dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:675)
    	at dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:670)
    	at dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:7073)
    	at dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:7073)
    	at dotty.tools.dotc.core.Types$TypeAccumulator.op$proxy28$1(Types.scala:6983)
    	at dotty.tools.dotc.core.Types$TypeAccumulator.foldArgs$3(Types.scala:6983)
    	at dotty.tools.dotc.core.Types$TypeAccumulator.foldOver(Types.scala:6987)
    	at dotty.tools.dotc.core.Types$TypeTraverser.traverseChildren(Types.scala:7074)
    	at dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:692)
    	at dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:7073)
    	at dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:7073)
    	at dotty.tools.dotc.typer.ProtoTypes$ViewProto.fold(ProtoTypes.scala:699)
    	at dotty.tools.dotc.core.Types$TypeAccumulator.foldOver(Types.scala:7036)
    	at dotty.tools.dotc.core.Types$TypeTraverser.traverseChildren(Types.scala:7074)
    	at dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:692)
    	at dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.apply(Implicits.scala:698)
    	at dotty.tools.dotc.typer.ImplicitRunInfo.recur$1(Implicits.scala:774)
    	at dotty.tools.dotc.typer.ImplicitRunInfo.computeIScope(Implicits.scala:785)
    	at dotty.tools.dotc.typer.ImplicitRunInfo.$anonfun$1(Implicits.scala:869)
    	at dotty.tools.dotc.util.ReadOnlyMap.getOrElse(ReadOnlyMap.scala:23)
    	at dotty.tools.dotc.typer.ImplicitRunInfo.implicitScope(Implicits.scala:869)
    	at dotty.tools.dotc.typer.ImplicitRunInfo.implicitScope$(Implicits.scala:636)
    	at dotty.tools.dotc.Run.implicitScope(Run.scala:45)
    	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.implicitScope(Implicits.scala:1817)
    	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.searchImplicit(Implicits.scala:1696)
    	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.searchImplicit$$anonfun$3(Implicits.scala:1786)
    	at dotty.tools.dotc.typer.Implicits$SearchResult.recoverWith(Implicits.scala:434)
    	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.searchImplicit(Implicits.scala:1772)
    	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.bestImplicit(Implicits.scala:1813)
    	at dotty.tools.dotc.typer.Implicits.inferImplicit(Implicits.scala:1126)
    	at dotty.tools.dotc.typer.Implicits.inferImplicit$(Implicits.scala:880)
    	at dotty.tools.dotc.typer.Typer.inferImplicit(Typer.scala:163)
    	at dotty.tools.dotc.typer.Implicits.inferView(Implicits.scala:920)
    	at dotty.tools.dotc.typer.Implicits.inferView$(Implicits.scala:880)
    	at dotty.tools.dotc.typer.Typer.inferView(Typer.scala:163)
    	at dotty.tools.dotc.typer.Typer.adaptToSubType$1(Typer.scala:4969)
    	at dotty.tools.dotc.typer.Typer.adaptNoArgsOther$1(Typer.scala:4787)
    	at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:4867)
    	at dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:5131)
    	at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:4347)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3965)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3970)
    	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:4082)
    	at dotty.tools.dotc.typer.Typer.$anonfun$69(Typer.scala:3130)
    	at dotty.tools.dotc.inlines.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:256)
    	at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:3130)
    	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3766)
    	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3877)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3965)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3970)
    	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3992)
    	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:4039)
    	at dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:1537)
    	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1541)
    	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3792)
    	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3878)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3965)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3970)
    	at dotty.tools.dotc.typer.Typer.typedFunctionValue(Typer.scala:2072)
    	at dotty.tools.dotc.typer.Typer.makeContextualFunction(Typer.scala:3941)
    	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3875)
    	at dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$7(ProtoTypes.scala:563)
    	at dotty.tools.dotc.typer.ProtoTypes$FunProto.cacheTypedArg(ProtoTypes.scala:486)
    	at dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArg(ProtoTypes.scala:565)
    	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:1179)
    	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:1179)
    	at dotty.tools.dotc.typer.Applications$Application.addTyped$1(Applications.scala:750)
    	at dotty.tools.dotc.typer.Applications$Application.matchArgs(Applications.scala:836)
    	at dotty.tools.dotc.typer.Applications$Application.init(Applications.scala:629)
    	at dotty.tools.dotc.typer.Applications$TypedApply.<init>(Applications.scala:1026)
    	at dotty.tools.dotc.typer.Applications$ApplyToUntyped.<init>(Applications.scala:1178)
    	at dotty.tools.dotc.typer.Applications.ApplyTo(Applications.scala:1485)
    	at dotty.tools.dotc.typer.Applications.ApplyTo$(Applications.scala:489)
    	at dotty.tools.dotc.typer.Typer.ApplyTo(Typer.scala:163)
    	at dotty.tools.dotc.typer.Applications.simpleApply$1(Applications.scala:1275)
    	at dotty.tools.dotc.typer.Applications.$anonfun$17(Applications.scala:1391)
    	at dotty.tools.dotc.typer.Typer$.tryEither(Typer.scala:136)
    	at dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:1410)
    	at dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1459)
    	at dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:489)
    	at dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:163)
    	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3784)
    	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3878)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3965)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3970)
    	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:4082)
    	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1544)
    	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3792)
    	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3878)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3965)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3970)
    	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:4019)
    	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:4039)
    	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:3433)
    	at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:3772)
    	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3776)
    	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3877)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3965)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3970)
    	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3992)
    	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:4039)
    	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:3568)
    	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3818)
    	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3878)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3965)
    	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3970)
    	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:4082)
    	at dotty.tools.dotc.typer.TyperPhase.typeCheck$$anonfun$1(TyperPhase.scala:47)
    	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    	at dotty.tools.dotc.core.Phases$Phase.monitor(Phases.scala:533)
    	at dotty.tools.dotc.typer.TyperPhase.typeCheck(TyperPhase.scala:53)
    	at dotty.tools.dotc.typer.TyperPhase.$anonfun$4(TyperPhase.scala:99)
    	at scala.collection.Iterator$$anon$6.hasNext(Iterator.scala:495)
    	at scala.collection.Iterator$$anon$9.hasNext(Iterator.scala:599)
    	at scala.collection.immutable.List.prependedAll(List.scala:156)
    	at scala.collection.immutable.List$.from(List.scala:682)
    	at scala.collection.immutable.List$.from(List.scala:682)
    	at scala.collection.IterableOps$WithFilter.map(Iterable.scala:911)
    	at dotty.tools.dotc.typer.TyperPhase.runOn(TyperPhase.scala:98)
    	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:380)
    	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1324)
    	at dotty.tools.dotc.Run.runPhases$1(Run.scala:373)
    	at dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$2(Run.scala:420)
    	at dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$adapted$1(Run.scala:420)
    	at scala.Function0.apply$mcV$sp(Function0.scala:42)
    	at dotty.tools.dotc.Run.showProgress(Run.scala:482)
    	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:420)
    	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:432)
    	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
    	at dotty.tools.dotc.Run.compileUnits(Run.scala:432)
    	at dotty.tools.dotc.Run.compileSources(Run.scala:319)
    	at dotty.tools.dotc.Run.compile(Run.scala:304)
    	at dotty.tools.dotc.Driver.doCompile(Driver.scala:38)
    	at dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:141)
    	at dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
    	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
    	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:196)
    	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
    	at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:252)
    	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:186)
    	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:166)
    	at sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:241)
    	at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:166)
    	at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:214)
    	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:542)
    	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:542)
    	at sbt.internal.inc.Incremental$.$anonfun$apply$3(Incremental.scala:182)
    	at sbt.internal.inc.Incremental$.$anonfun$apply$3$adapted(Incremental.scala:180)
    	at sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:458)
    	at sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:117)
    	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
    	at sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
    	at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
    	at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:413)
    	at sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:500)
    	at sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:400)
    	at sbt.internal.inc.Incremental$.apply(Incremental.scala:208)
    	at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:542)
    	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:496)
    	at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
    	at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:433)
    	at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
    	at sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2443)
    	at sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2393)
    	at sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:41)
    	at sbt.internal.io.Retry$.apply(Retry.scala:47)
    	at sbt.internal.io.Retry$.apply(Retry.scala:29)
    	at sbt.internal.io.Retry$.apply(Retry.scala:24)
    	at sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:41)
    	at sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2391)
    	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
    	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:63)
    	at sbt.std.Transform$$anon$4.work(Transform.scala:69)
    	at sbt.Execute.$anonfun$submit$2(Execute.scala:283)
    	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
    	at sbt.Execute.work(Execute.scala:292)
    	at sbt.Execute.$anonfun$submit$1(Execute.scala:283)
    	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
    	at sbt.CompletionService$$anon$2.call(CompletionService.scala:65)
    	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
    	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    	at java.base/java.lang.Thread.run(Thread.java:840)
```
</details>

In my code, I have a type defined like this:

```scala
final class TreeChainMap[A, B](
    private val tree: Evermean[A, B]
)(using PartialOrder[A]):
  def union[D <: AnyKind](other: (h: HKT[D]) ?=> h.Ar1[(A, B)] | h.Ar2[A, B])(
      using Foldable2[D]
  )(using h: HKT[D]): Maybe[TreeChainMap[A, B]] = ???

//elsewhere
    val x: TreeChainMap[Int, Int] =
      TreeChainMap.empty[Int, Int]

    x.union(HKT(x)) //typechecks and compiles
    x.union(x) //x doesn't fit the type signature on its own. Should cause a compilation error, but rather crashes the compiler
```
As noted, when I pass in a valid type to union, the compiler compiles the code. When I don't, the compiler crashes rather than giving a compilation error.

--- a/compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala
+++ b/compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala
@@ -955,6 +955,13 @@ object ProtoTypes {
     }
   }
 
+  def containsTypeParamRef(tp: Type)(using Context): Boolean =
+    tp.stripped match {
+      case tr: TypeParamRef => true
+      case AppliedType(_, args) => args.exists(containsTypeParamRef)
+      case _ => false
+    }
+
   /** Approximate occurrences of parameter types and uninstantiated typevars
    *  by wildcard types.
    */
@@ -962,7 +969,9 @@ object ProtoTypes {
     tp match {
     case tp: NamedType => // default case, inlined for speed
       val isPatternBoundTypeRef = tp.isInstanceOf[TypeRef] && tp.symbol.isPatternBound
+      val isImplicitTermRef = tp.isInstanceOf[TermRef] && tp.symbol.maybeOwner.isAnonymousFunction && tp.symbol.isParamOrAccessor && tp.symbol.is(Flags.Given) && containsTypeParamRef(tp.superType)
       if (isPatternBoundTypeRef) WildcardType(tp.underlying.bounds)
+      else if (isImplicitTermRef) wildApprox(tp.underlying, theMap, seen, internal)
       else if (tp.symbol.isStatic || (tp.prefix `eq` NoPrefix)) tp
       else tp.derivedSelect(wildApprox(tp.prefix, theMap, seen, internal))
     case tp @ AppliedType(tycon, args) =>

--- a/tests/neg/i25500.scala
+++ b/tests/neg/i25500.scala
@@ -1,0 +1,7 @@
+trait A[B[_]]:
+  type C
+
+def apply[D[_]](other: (a: A[D]) ?=> a.C): Any = ???
+
+def main =
+  apply(()) // error

--- a/tests/pos/i25500.scala
+++ b/tests/pos/i25500.scala
@@ -1,0 +1,13 @@
+trait A[B[_]]:
+  type C
+
+object A:
+  given A[Option]:
+    type C = Option[Int]
+
+  def apply[E[_], F](e: E[F])(using a: A[E]): a.C = ???
+
+def apply[D[_]](other: (a: A[D]) ?=> a.C): Any = ???
+
+def main =
+  apply(A(Option(5)))

